src/proxies/forwarded.rs

```diff
@@ -0,0 +1,647 @@
+        }
+    }
+
+    /// Parse a borrowed Forwarded from the historical
+    /// non-standardized x-forwarded-{for,by,proto} headers, without
+    /// support for the Forwarded header.
+    ///
+    /// # Examples
+    /// ```rust
+    /// # use http_types::{proxies::Forwarded, Method::Get, Request, Url};
+    /// let mut request = Request::new(Get, Url::parse("http://_/").unwrap());
+    /// request.insert_header("X-Forwarded-For", "192.0.2.43, 2001:db8:cafe::17");
+    /// let forwarded = Forwarded::from_headers(&request).unwrap().unwrap();
+    /// assert_eq!(forwarded.forwarded_for(), vec!["192.0.2.43", "[2001:db8:cafe::17]"]);
+    /// ```
+    /// ```rust
+    /// # use http_types::{proxies::Forwarded, Method::Get, Request, Url};
+    /// let mut request = Request::new(Get, Url::parse("http://_/").unwrap());
+    /// request.insert_header(
+    ///     "Forwarded",
+    ///     r#"for=192.0.2.43, for="[2001:db8:cafe::17]", for=unknown;proto=https"#
+    /// );
+    /// assert!(Forwarded::from_x_headers(&request).unwrap().is_none());
+    /// ```
+    pub fn from_x_headers(headers: &'a impl AsRef<Headers>) -> Result<Option<Self>, ParseError> {
+        let headers = headers.as_ref();
+
+        let forwarded_for: Vec<Cow<'a, str>> = headers
+            .get(X_FORWARDED_FOR)
+            .map(|hv| {
+                hv.as_str()
+                    .split(',')
+                    .map(|v| {
+                        let v = v.trim();
+                        match v.parse::<IpAddr>().ok() {
+                            Some(IpAddr::V6(v6)) => Cow::Owned(format!(r#"[{}]"#, v6)),
+                            _ => Cow::Borrowed(v),
+                        }
+                    })
+                    .collect()
+            })
+            .unwrap_or_default();
+
+        let by = headers
+            .get(X_FORWARDED_BY)
+            .map(|hv| Cow::Borrowed(hv.as_str()));
+
+        let proto = headers
+            .get(X_FORWARDED_PROTO)
+            .map(|p| Cow::Borrowed(p.as_str()));
+
+        if !forwarded_for.is_empty() || by.is_some() || proto.is_some() {
+            Ok(Some(Self {
+                forwarded_for,
+                by,
+                proto,
+                host: None,
+            }))
+        } else {
+            Ok(None)
+        }
+    }
+
+    /// parse a &str into a borrowed Forwarded
+    pub fn parse(input: &'a str) -> Result<Self, ParseError> {
+        let mut input = input;
+        let mut forwarded = Forwarded::new();
+
+        if starts_with_ignore_case("for=", input) {
+            input = forwarded.parse_for(input)?;
+        }
+
+        loop {
+            match input.chars().next() {
+                Some(';') => {
+                    input = &input[1..];
+                }
+                None => return Ok(forwarded),
+                _ => return Err(ParseError::new("unexpected character after forwarded-pair")),
+            }
+
+            input = forwarded.parse_forwarded_pair(input)?;
```

Comment:
No, because we don't start with a `;`.

this is valid: `Forwarded: proto=https`

Reply:
On further thought, I believe you're right that it would be identical to move it into the Some(';'), and also that represents a bug. I think we do not correctly parse a `Forwarded: proto=https` header

Reply:
Fwiw: Presently this line only gets hit if the `Some(";")` branch is taken, all other options return beforehand.

Reply:
Addressed in 254a405, this was in fact a bug. Thanks!